### PR TITLE
Improve Android user experience

### PR DESCRIPTION
**Pull request by dg0yt**
- Minor build system adjustments
- Fixes for icon and dock widget issues
- Relaxed permission handling on the home screen

Resolves GH-1293.
Closes GH-787.

### CODE REVIEW

**Comment by pkturner** [comment]
> Resolves GH-1292.
This looks unintentional, as #1292 is unrelated to this PR.

**Comment by dg0yt** [comment]
> > Resolves [GH-1292](https://github.com/OpenOrienteering/mapper/pull/1292).
> 
> This looks unintentional, as #1292 is unrelated to this PR.

GH-1293, of course. Done.

File: src/core/storage_location.cpp [hunk 1]
```diff
@@ -59,6 +59,11 @@ bool storageAccessGranted()
 	return AppPermissions::checkPermission(AppPermissions::StorageAccess) == AppPermissions::Granted;
 }
 
+/**
+ * Tells Android to update its media scanner cache.
+ * 
+ * This helps to avoid corrupted files when transfering files with MTP.
+ */
```
Comment (dg0yt) [inline src/core/storage_location.cpp]: Already documented in the header file. I will remove this commit.

File: src/gui/widgets/mapper_proxystyle.cpp [hunk 2]
```diff
@@ -36,6 +37,29 @@
...
+namespace  {
+
+/**
+ * This function recognizes dock widget related widgets by the class name
+ * matching "DockWidget".
+ * 
+ * This function helps to customize the style for classes like
+ * "QDockWidgetTitleButton" or "MapEditorDockWidget".
+ */
+bool isDockWidgetRelated(const QWidget* widget)
```
Comment (dg0yt) [inline src/gui/widgets/mapper_proxystyle.cpp]: This causes an unused-function warning on all systems but Android. I will have to add `#ifdef Q_OS_ANDROID`

Comment (dg0yt) [reply src/gui/widgets/mapper_proxystyle.cpp]: Added `Q_DECL_UNUSED` instead.

**Comment by lpechacek** [comment]
Will do my best to review until end of week.

**Comment by dg0yt** [comment]
Android foreground service is coming, too: 
https://github.com/OpenOrienteering/mapper/compare/dg0yt/android-service

However, it cannot pass CI until the superbuild update to compile SDK >= 26 is integrated. 
https://github.com/OpenOrienteering/superbuild/pull/37

File: src/gui/main_window.cpp [hunk 3]
```diff
@@ -498,7 +499,19 @@ void MainWindow::setHasUnsavedChanges(bool value)
 		has_unsaved_changes = value;
 		setAutosaveNeeded(has_unsaved_changes && !has_autosave_conflict);
 	}
+	else
+	{
```
Comment (dg0yt) [inline src/gui/main_window.cpp]: The following needs more work.

**Comment by dg0yt** [comment]
Thanks a lot for the review.

> Please make code self-modification in "Android: Fix BuildConfig and R access" more visible to a future developer trying to debug that part. [...] Example: […]

Well, this really far from perfect. Guess how long it took me to realize why R and BuildConfig wouldn't work out of the box.
I did consider something like you propose, but then you would need to remember to add these headers to new classes. However, from an Android developer point of view, you would expect them to be available right in the app's namespace, without explicit import. 
So I really would like to come to the proper solution which seems to be "flavours" in gradle, completely eliminating this hack. But it is out of scope for the current PR. The interplay of gradle, androiddeployqt, local setups and CI deserve its own share of time.

> Button for hiding the top Mapper menu bar changes size. It's smaller when the menu is hidden.

Good catch.

> Mapper crashed three times during testing. Android system log contains following trace for all three crashes: [...] 

Maybe this is unrelated to our source code. but a bug in Android >= P, cf. 
https://stackoverflow.com/questions/51956971/illegalstateexception-of-toast-view-on-android-p-preview
even with a google issue link. I can try to create a similar workaround.
Similar:
https://stackoverflow.com/questions/54624116/android-java-lang-illegalstateexception-in-toast-handleshow

**Comment by dg0yt** [comment]
> So I really would like to come to the proper solution which seems to be "flavours" in gradle, completely eliminating this hack.

However, androiddeployqt and/or Qt Creator are not ready for changing the app ID in build.gradle:  
https://bugreports.qt.io/browse/QTCREATORBUG-20802

**Comment by dg0yt** [comment]
>     * Please provide more context for the GitHub issue references in commit messages. The message should be self-contained, therefore understandable  without having to connect to Internet (I do most of my Mapper work offline on train). Example: `This is expected to resolve #576.` -> `This is expected to resolve GH-576 ("Track recording when screen is turned off").`

Done.

>     * Please make code self-modification in "Android: Fix BuildConfig and R access" more visible to a future developer trying to debug that part. Efforts to figure out why there are differences between the source code repository and compile-time sources typically take half day to complete with facepalm and swear words at the end. I'd welcome if code added by the `sed` script were present in the source file and commented out with a remark that a build script activates this code during build. Example: [...]

Created #1306 instead.

>     * Button for hiding the top Mapper menu bar changes size. It's smaller when the menu is hidden.

Done.

>     * Mapper crashed three times during testing. Android system log contains following trace for all three crashes: [...]

Created #1305 instead.

In addition, fixed a crash with large buttons on tiny screens.

File: src/gui/widgets/editor_settings_page.cpp [hunk 4]
```diff
@@ -44,6 +45,12 @@ EditorSettingsPage::EditorSettingsPage(QWidget* parent)
 {
 	auto layout = new QFormLayout(this);
 	
+	if (MainWindow::mobileMode())
+	{
+		button_size = Util::SpinBox::create(1, 3.0, 26.0, tr("mm", "millimeters"), 0.1);
```
Comment (dg0yt) [inline src/gui/widgets/editor_settings_page.cpp]: @lpechacek How about making this an integer, similar to the following 'Symbol icon size'? It is always a little bit difficult to edit this setting. At least I would like to change the step size from 0.1 to 0.5.

**Pull request merged**